feat: add benchmarks and performance checks on PRs

ajnavarro opened the pull request:
# Description

Add benchmark monitoring to keep track of improved/decreased performance. We also added checks to run fast benchmarks on every PR.

it closes #689 

Still a WIP:
- [ ] Use a GitHub self-hosted runner to keep benchmark results consistent
- [x] Add a README.md inside the `.benchmarks` folder to help people add new benchmarks to graphs and PR checks.
- [x] Remove -short flag when generating the graphs

## Contributors Checklist

- [X] Added new tests, or not needed, or not feasible
- [X] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [X] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [X] Added references to related issues and PRs
- [X] Provided any useful hints for running manual tests

![screenshot-localhost_8080-2023 06 08-15_44_09](https://github.com/gnolang/gno/assets/1196465/73036e44-c1e2-484b-bc06-1f3b867b4161)

moul commented on .benchmarks/gobenchdata-checks.yml:
```diff
@@ -0,0 +1,25 @@
+      benchmarks: []
+      diff: (current.NsPerOp - base.NsPerOp) / base.NsPerOp * 100
+      thresholds:
+        max: 10
```
what it means, and what happens if we hit this?

ajnavarro replied:
example check output report: https://github.com/ajnavarro/gno/actions/runs/5280312972/jobs/9553479785#step:4:63

moul commented on .benchmarks/gobenchdata-checks.yml:
```diff
@@ -0,0 +1,25 @@
+checks:
+    - name: VM checks
+      description: |-
+        It checks performance regressions on VM package.
```
```suggestion
        It checks performance regressions on VM full test suite.
```

or something else, but 'package' is invalid.

moul commented on .benchmarks/gobenchdata-web.yml:
```diff
@@ -0,0 +1,45 @@
+  - name: VM Benchmarks
+    description: |
+      Benchmarks testing Gno VM
+    charts:
+      - name: Type assertion benchmarks (gnovm)
+        benchmarks: [ 'TypeAssertion.' ]
+        package: github.com\/gnolang\/gno\/gnovm\/pkg\/gnolang
+      - name: Reflect benchmarks (gnovm)
+        benchmarks: [ 'Reflect.' ]
+        package: github.com\/gnolang\/gno\/gnovm\/pkg\/gnolang
+      - name: Equality benchmarks (gnovm)
+        benchmarks: [ '.Equality' ]
+        package: github.com\/gnolang\/gno\/gnovm\/pkg\/gnolang
+      - name: LoopyMain benchmarks (gnovm)
+        benchmarks: [ '.LoopyMain' ]
+        package: github.com\/gnolang\/gno\/gnovm\/pkg\/gnolang
+      - name: Mempool (tm2)
+        package: ^github.com/gnolang/gno/tm2/pkg/bft/mempool$
+      - name: State (tm2)
+        package: ^github.com/gnolang/gno/tm2/pkg/bft/state$
+      - name: Types (tm2)
+        package: ^github.com/gnolang/gno/tm2/pkg/bft/types$
+      - name: WAL (tm2)
+        package: ^github.com/gnolang/gno/tm2/pkg/bft/wal$
+      - name: Bcrypt (tm2)
+        package: ^github.com/gnolang/gno/tm2/pkg/crypto/bcrypt$
+      - name: ed25519 (tm2)
+        package: ^github.com/gnolang/gno/tm2/pkg/crypto/ed25519$
+      - name: Merkle (tm2)
+        package: ^github.com/gnolang/gno/tm2/pkg/crypto/merkle$
+      - name: secp256k1 (tm2)
+        package: ^github.com/gnolang/gno/tm2/pkg/crypto/secp256k1$
+      - name: iavl (tm2)
+        package: ^github.com/gnolang/gno/tm2/pkg/iavl.
+      - name: random (tm2)
+        package: ^github.com/gnolang/gno/tm2/pkg/random$
+      - name: auth (tm2)
+        package: ^github.com/gnolang/gno/tm2/pkg/sdk/auth$
+      - name: cache (tm2)
+        package: ^github.com/gnolang/gno/tm2/pkg/store/cache$
```
we can probably generate this dynamically

moul commented on .github/workflows/benchmark-check.yml:
```diff
@@ -0,0 +1,22 @@
+    - uses: actions/setup-go@v2
+      with: { go-version: "1.20" }
+    - name: gobenchdata publish
+      run: go run go.bobheadxi.dev/gobenchdata@v1 action
```
can you add a go install step before; just to make sure that we've more comprehensible error handling and sense of speed for install and run.

moul commented:
wow, that looks powerful, and I like how it replaces an infra with git 👍

moul commented on .github/workflows/benchmark-publish.yml:
```diff
@@ -0,0 +1,23 @@
+name: benchmarks publish
```
can this be merged to have a single github action; maybe by merging the `on:` triggers and adding a `if push + branch=master` condition?

ajnavarro replied:
hmm, they are two different things, one is adding information about performance deltas on the PR, and the other is generating the dashboard with all the benchmark results. I can merge them if it clarifies what benchmark action does.

moul commented on .github/workflows/benchmark-publish.yml:
```diff
@@ -0,0 +1,23 @@
+      env:
+        GITHUB_TOKEN: ${{ secrets.ACCESS_TOKEN }}
+        INPUT_PRUNE_COUNT: 30
+        INPUT_GO_TEST_FLAGS: -short -run=^$ # TODO: remove -short flag
```
if too long, let's keep a short test for PRs and a long one once a day against master.

ajnavarro replied:
done.

ajnavarro commented:
Until I have a self-hosted runner managed by the company and a token with write permissions, I managed to do it on my own on a fork and using my home server to run the benchmarks:

https://ajnavarro.com/gno-benchmarks/

I'm having some problems with the second run. I opened an issue: https://github.com/bobheadxi/gobenchdata/issues/76

ajnavarro commented:
Everything is tested and working on my side using a fork, we only need an access token with write permissions to write into benchmarks gh-pages branch all the results, and a stateless self-hosted runner to have consistent results every time we execute the tests.

CC @moul @albttx

zivkovicmilos commented on tm2/pkg/std/coin_benchmark_test.go:
```diff
@@ -7,6 +7,7 @@ import (
 )
 
 func BenchmarkCoinsAdditionIntersect(b *testing.B) {
+	b.Skip("TODO: panicking benchmark")
```
Can we open up an issue for these?

ajnavarro replied:
https://github.com/gnolang/gno/issues/907

zivkovicmilos commented on tm2/pkg/iavl/benchmarks/bench_test.go:
```diff
@@ -181,6 +181,7 @@ type benchmark struct {
 }
 
 func BenchmarkMedium(b *testing.B) {
+	b.Skip("TODO: benchmark panicking")
```
Same issue for these as well

zivkovicmilos commented on tm2/pkg/crypto/secp256k1/bench_test.go:
```diff
@@ -9,6 +9,8 @@ import (
 )
 
 func BenchmarkKeyGeneration(b *testing.B) {
+	b.Skip("too slow (more than 30 mins)")
```
See https://github.com/gnolang/gno/pull/670

ajnavarro replied:
Does that mean that it is still too slow?

zivkovicmilos replied:
Yeah, it's super slow for even small loop calls (ex. like 100 calls), let alone these benchmark limits

zivkovicmilos commented on tm2/pkg/bft/wal/wal_test.go:
```diff
@@ -262,21 +262,26 @@ func BenchmarkWalRead10KB(b *testing.B) {
 }
 
 func BenchmarkWalRead100KB(b *testing.B) {
+	b.Skip("TODO: benchmark failing")
```
Can these be a separate issue, since they relate to the WAL?

ajnavarro replied:
https://github.com/gnolang/gno/issues/910

zivkovicmilos commented on tm2/pkg/amino/benchmark_test.go:
```diff
@@ -37,12 +44,17 @@ func BenchmarkBinaryPBBindings(b *testing.B) {
 			_benchmarkBinary(b, cdc, rt, "binary_pb", true)
 		})
+
+		// TODO: fix nil pointer error
```
Leftover TODO?

ajnavarro replied:
It was actually failing with a nil pointer

zivkovicmilos commented on tm2/pkg/amino/benchmark_test.go:
```diff
@@ -6,14 +6,19 @@ import (
 
-	"github.com/gnolang/gno/tm2/pkg/amino"
-	"github.com/gnolang/gno/tm2/pkg/amino/tests"
 	fuzz "github.com/google/gofuzz"
```
Just curious, why do we use the gofuzz package, since Go supports fuzzing out of the box as of 1.18? @moul

ajnavarro replied:
I think because the code is older than the introduction of Fuzzing in Go

zivkovicmilos commented on .github/pull_request_template.md:
```diff
@@ -12,6 +12,7 @@ Please provide a detailed description of the changes made in this pull request.
 - [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
 - [ ] Added references to related issues and PRs
 - [ ] Provided any useful hints for running manual tests
+- [ ] Add new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](../.benchmarks/README.md).
```
It should probably be `Added`, to follow other items

zivkovicmilos commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,33 @@
+      uses: actions/checkout@v2
+      with: 
+        fetch-depth: 1
+    - uses: actions/setup-go@v2
```
You can bump these actions to version 4: `actions/setup-go@v4`

zivkovicmilos commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,33 @@
+    runs-on: ubuntu-latest
+    steps:
+    - name: checkout
+      uses: actions/checkout@v2
```
You can bump these actions to version 3: `actions/checkout@v3`

zivkovicmilos commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,33 @@
+        fetch-depth: 1
+    - uses: actions/setup-go@v2
+      with: 
+        go-version: "1.20"
```
What do you think about doing matrix go versions? (1.19 / 1.20)

ajnavarro replied:
Because it will execute all benchmarks twice. I think we should use the latest available go version here to use all the performance improvements we can from Go itself. Why do you think it might be necessary to run using two Go versions?

zivkovicmilos replied:
Ah, got it - no specific reason for needing to run it twice. Might be interesting to see if there is a bump between Go versions, but it's an unnecessary overhead now 🙃

zivkovicmilos commented on .benchmarks/README.md:
```diff
@@ -0,0 +1,32 @@
+# Benchmarks
+
+This folder is where benchmarks are configured to be added on the dashboard generated in [benchmarks](https://gnoland.github.io/benchmarks).
```
@moul @albttx
Can we host these at `benchmarks.gno.land`, or somewhere similar?

moul replied:
good idea, let's do it

zivkovicmilos reviewed (approved):
I like this effort very much - it's a start in the right direction for performance tracking

I've left minor comments, otherwise looks great 💯

ajnavarro commented:
As suggested by @moul , I changed to publish all benchmark results into a local branch, `gh-benchmarks` and added a process to copy that branch into benchmark's gh-pages branch here: https://github.com/gnolang/benchmarks/pull/7
